servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherCache.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.PublisherSource.Subscriber;
+import io.servicetalk.concurrent.PublisherSource.Subscription;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/**
+ * A cache of {@link Publisher}s. Keys must correctly implement Object::hashCode and Object::equals.
+ * The cache can be created with either multicast or replay configurations. Subscriptions
+ * are tracked by the cache and Publishers are removed when there are no more subscriptions.
+ *
+ * @param <K> a key type suitable for use as a Map key, that correctly implements hashCode/equals.
+ * @param <T> the type of the publisher that is cached.
+ */
+public final class PublisherCache<K, T> {
+    private final MulticastStrategy<T> multicastStrategy;
+    private final Map<K, Holder<T>> publisherCache;
+
+    PublisherCache(final MulticastStrategy<T> multicastStrategy) {
+        this.multicastStrategy = multicastStrategy;
+        this.publisherCache = new HashMap<>();
+    }
+
+    /**
+     * Create a new PublisherCache where the cached publishers must be configured with a multicast or replay operator
+     * by the multicastSupplier function.
+     *
+     * @param <K> a key type suitable for use as a {@link Map} key.
+     * @param <T> the type of the {@link Publisher} contained in the cache.
+     * @return a new PublisherCache that will not wrap cached values.
+     */
+    public static <K, T> PublisherCache<K, T> create() {
+        return new PublisherCache<>(MulticastStrategy.identity());
+    }
+
+    /**
+     * Create a new PublisherCache where the cached publishers will be configured for multicast for all
+     * consumers of a specific key.
+     *
+     * @param <K> a key type suitable for use as a {@link Map} key.
+     * @param <T> the type of the {@link Publisher} contained in the cache.
+     * @return a new PublisherCache that will wrap cached values with multicast operator.
+     */
+    public static <K, T> PublisherCache<K, T> multicast() {
```

Comment:
by exposing these static method we may need to add a new factory method here for each operator overload. did you consider instead exposing just a `Function<>` (or similar) so folks can apply the variant they want? some risks maybe:
- folks could apply operators that don't obey the assumptions (we could clarify the constraints the operator must abide by ... allows for multiple subscribers, cancels upstream only after no subscribers present, no async cancel processing as the synchronization here depends upon it, ..)

Reply:
I did consider this. I'm open to removing these, the `create()` method just above is essentially what you describe, the user would specify the `Publisher` configured however they desire in the function on `get`. 

I had initially built this without the static methods and a constructor that took a function which would be used to supply a new Publisher on a cache miss. I decided to move this function to the `get` method as it emulated how I might expect to use a cache, ex: I might not want a function from name -> Publisher but rather I would prefer a closure that would allow me to use the context at hand to instantiate the new object.

For an initial API I don't have any problem deferring to your suggestion as we learn how this ultimately ends up being used.